components/suggest/src/rs.rs

```diff
@@ -436,6 +436,19 @@ impl DownloadedSuggestionCommonDetails {
             },
         )
     }
+
+    /// Get the full keywords as a single string
+    pub fn full_keywords_joined(&self) -> String {
+        let parts: HashSet<_> = self.full_keywords.iter().map(|(s, _)| s.as_str()).collect();
+        let mut result = String::new();
+        for (i, part) in parts.into_iter().enumerate() {
+            if i == 0 {
```

Comment:
This should be `i > 0`, right? And IIUC you're using a hash set to discard duplicate full keywords, but that's not the problem, and I'd be surprised if AMP included duplicate full keywords for one suggestion. The problem is that many of a suggestion's full keywords can contain the same substring, e.g. "amazon". So you'd need to `split_whitespace()` on each full keyword string and then collect each resulting word into a hash set. That should work.

You know, it might be worth preprocessing this and including the joined full keywords string in the data itself. I bet it wouldn't be too hard to modify our AMP upload script, and it probably wouldn't increase the dataset size by too much.

Reply:
Ah my `i > 0` comment is outdated, thanks.